### ISSUE TITLE
Code shortage

### ISSUE THREAD

**Issue by mjdf2568** [jonggyujang0123/QoS-UARAPC#1]
Hello, can you upload the code for how to train the models section?

**Comment by jonggyujang0123** [comment]
Hi, I found some version of the `RL` training code (you may run it `~_v4` to `~`.
I am sorry that I am no longer implement it due to GPU issue.. 
```python
# -*- coding: utf-8 -*-

import numpy as np
import random
import tensorflow as tf
#import scipy.misc1
from environment_v4 import ENV_net
from actor_critic_net_v4 import USL, USL_replay, Actor, Actor_USL, Critic, DDPG, PER, GaussianNoise
import scipy.io as sio
import os
os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"   # see issue #152
os.environ["CUDA_VISIBLE_DEVICES"]="0"



Is_resume = False

save_file_UA = './UA_model/UA_model'
save_file_PC = './PC_model/PC_model'

UE = 40
BS = 4


Max_ep = 8000



# SNR_mat = sio.loadmat('../../../DATA_GENERATOR_8BS/SNR_mat_v10.mat')['SNR'][:,:,:,0:BS]
SNR_mat = 200 * sio.loadmat('../../DATA_GENERATOR_8BS/SNR_test_v12.mat')['SNR'][:,:,:,0:BS]
# SNR_mat = SNR_mat[0:3000,:,:,:]

env = ENV_net(SNR_mat, UE, BS)
Max_step = SNR_mat.shape[1]-1


config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)

PC_agent = DDPG(scope = 'PC_net',
                sess=sess,
                BS = BS,
                UE = UE,
                Actor = Actor(env.action_size_PC, thr = UE * BS, scope = 'PC_net_DDPG_Actor'),
                Critic = Critic(env.reward_size_PC, BS, UE, scope = 'PC_net_DDPG_Critic'),
                Actor_target = Actor(env.action_size_PC, thr = UE * BS, scope = 'PC_net_DDPG_Actor_target'),
                Critic_target = Critic(env.reward_size_PC, BS, UE, scope = 'PC_net_DDPG_Critic_target'),
                OUNoise = GaussianNoise(env.action_size_PC,epsilon_init = 0.99,epsilon_end=0.3),
                replay_buffer = PER(buffer_size =int(2e5), alpha = 0.7, epsilon_per = 0.1, beta = 0.4), 
                state_size = env.state_size_PC, 
                action_size = env.action_size_PC, 
                reward_size = env.reward_size_PC,
                gamma = 0.9, 
                lr_actor = 3e-7, 
                lr_critic = 1e-5, 
                batch_size = 2048,
                tau = 0.003,
                is_tanh = True
                )

UA_agent = USL(scope = 'UA_net',
                sess= sess,
                BS = BS,
                UE = UE,
                Actor = Actor_USL(env.action_size_UA, scope = 'UA_net'),
                replay_buffer = USL_replay(buffer_size = 100000), 
                state_size = env.state_size_UA, 
                action_size = env.action_size_UA, 
                lr_actor = 3e-5,
                batch_size = 1024,
                alpha_init= 0.03
                )

init = tf.global_variables_initializer()
sess.run(init )
    
saver1 = tf.train.Saver(var_list=tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope='UA'))
saver1.restore(sess, save_file_UA)
saver2 = tf.train.Saver(var_list=tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope='PC'))




reward_list_PC = list()
sr_list = list()
qos_list = list()
loss_list = list()
BBB=0
AAA=0
PC_start = 1000
for env.episode in range(Max_ep):
    sr_wo_noise = list()
    reward_ep_PC = 0
    env.step = 0
    env.reset()
    PC_agent.noise.reset()
    PC_agent.noise.sigma =np.clip( 3.0 * (8000-env.episode)/7000, 0.1, 3.0) 
    loss_set_PC = list()
    QoS_prob = list()
    for env.step in range(Max_step):
        if env.episode < 1000:
            #env.action = np.random.uniform(-1, 1,size = [1,BS])
            env.action = PC_agent.forward_noise_action(env.state_PC, 30).copy()
        else:
            env.action = PC_agent.forward_noise_action(env.state_PC, env.step).copy()
        env.get_state_UA()
        env.action_UA = UA_agent.forward_action(env.state_UA).copy()
        env.proceed_master()
        QoS_prob.append(env.QoS_unsatisfactory.copy())
        sr_wo_noise.append(env.sum_rate.copy())
        PC_agent.add_exp(env.state_PC.copy(), env.state_next.copy(), env.action.copy(), env.reward.copy())
        reward_ep_PC = reward_ep_PC + env.reward
        env.state_PC = env.state_next + 0.0
    if len(PC_agent.replay_buffer.buffer) > 30000:
        _ = PC_agent.train()
    loss_np = PC_agent.forward_loss(env.state_PC, env.state_next, env.action, env.reward)
    loss_set_PC.append(loss_np.copy())
    qos_list.append(np.mean(QoS_prob))
    reward_list_PC.append(np.sum(reward_ep_PC[0,:])/Max_step)
    sr_list.append(np.mean(sr_wo_noise))
    loss_list.append(loss_set_PC)
    print('%i-th episodes PC_reward: %F, PC_loss: %f, QoS : %f' %(env.episode, np.sum(reward_ep_PC[0,:])/Max_step,np.mean(loss_set_PC),np.mean(QoS_prob)))
    if env.episode % 30 == 0 : 
        print(np.mean(sr_list[-300::]),np.mean(qos_list[-300::]))
    if env.episode%500==499:
        saver2.save(sess, save_file_PC)

sio.savemat('PC_reward_list.mat',{'data' : np.array(reward_list_PC)})
sio.savemat('PC_sr_list.mat',{'data' : np.array(sr_list)})
sio.savemat('PC_loss_list.mat',{'data' : np.array(loss_list)})
sio.savemat('PC_QoS_list.mat',{'data' : np.array(qos_list)})    
            
        
#    if Is_train == True and episode%100==0:
#        sio.savemat('reward_data_training.mat',{'data_reward' : data_reward})
#        sio.savemat('loss_data_training.mat',{'data_loss' : data_loss})
#        with open(save_file, 'wb') as file:
#            with open(save_file+'epi.txt', 'wb') as file:
#                pickle.dump(replay_buffer,file)
```

**Comment by jonggyujang0123** [comment]
This is code for unsupervised learning
```python

import numpy as np
import random
import tensorflow as tf
#import scipy.misc1
from environment_v4 import ENV_net
from actor_critic_net_v4 import USL, USL_replay, Actor, Actor_USL, Critic, DDPG, PER, GaussianNoise
import scipy.io as sio
import os
os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"   # see issue #152
os.environ["CUDA_VISIBLE_DEVICES"]="1"



Is_resume = False


UE = 40
BS = 4


Training_epi = 30000 
Training_epoch = 3000


SNR_mat = 200 * sio.loadmat('../../DATA_GENERATOR_8BS/SNR_test_v12.mat')['SNR'][:,:,:,0:BS]

env = ENV_net(SNR_mat, UE, BS)
Max_step = SNR_mat.shape[1]-1


config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)

UA_agent = USL(scope = 'UA_net',
                sess= sess,
                BS = BS,
                UE = UE,
                Actor = Actor_USL(env.action_size_UA, scope = 'UA_net'),
                replay_buffer = USL_replay(buffer_size = Training_epi*30), 
                state_size = env.state_size_UA, 
                action_size = env.action_size_UA, 
                lr_actor = 3e-5,
                batch_size = 2048,
                alpha_init= 0.03
                )


saver1 = tf.train.Saver(var_list=tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope='UA'))
# saver1.restore(sess, './model/UA_model')
# saver2 = tf.train.Saver(var_list=tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope='PC'))
# saver2.save(sess, './model/PC_model')
# saver2.restore(sess, './model/PC_model')

init = tf.global_variables_initializer()
if Is_resume == True:
    saver1.restore(sess, './UA_model/UA_model')
else: 
    sess.run(init )


for env.episode in range(Training_epi):
    sr_wo_noise = list()
    reward_ep_PC = 0
    reward_ep_UA = 0
    env.step = 0
    env.reset()
    if env.episode % 1000 == 0:
        print(env.episode)
    for env.step in range(Max_step):
        env.action = np.clip( np.array([[-0.4, -0.4, -0.4, 0.4]]) + np.random.uniform(-0.6,0.6, size=[1,BS]), -1, 1)
        env.get_state_UA()
        env.action_UA = np.random.uniform(0, 1, size=[1,2*BS]) #UA_agent.forward_action(env.state_UA).copy()
        UA_agent.add_exp(env.state_UA.copy(), env.Rate.copy(), env.QoS.copy(), env.Backhaul.copy())
        env.proceed_master(is_train = True)
        env.state_PC = env.state_next + 0.0
        
Loss_list = list()
Lag_list = list()
sr_list = list()
qos_list = list()

for epoch in range(Training_epoch):
    Loss, Lag, SR, QoS = UA_agent.train(is_print=True)
    Loss_list.append(Loss)
    Lag_list.append(Lag)
    sr_list.append(SR)
    qos_list.append(QoS)
    print('%i-th epochs UA_loss: %f' %(epoch, Loss**0.5))
    if epoch%500==499:
        saver1.save(sess, './UA_model/UA_model')


sio.savemat('UA_lag_list.mat',{'data' : np.array(Lag_list)})
sio.savemat('UA_Loss_list.mat',{'data' : np.array(Loss_list)})
sio.savemat('UA_sr_list.mat',{'data' : np.array(sr_list)})
sio.savemat('UA_qos_list.mat',{'data' : np.array(qos_list)})
            
        
#    if Is_train == True and episode%100==0:
#        sio.savemat('reward_data_training.mat',{'data_reward' : data_reward})
#        sio.savemat('loss_data_training.mat',{'data_loss' : data_loss})
#        with open(save_file, 'wb') as file:
#            with open(save_file+'epi.txt', 'wb') as file:
#                pickle.dump(replay_buffer,file)

**Comment by jonggyujang0123** [comment]
I am unsure if this is the final version, but you may reproduce the replica version of our results.

**Comment by mjdf2568** [comment]
Thank you very much for your reply！
